purchase_sale_stock_inter_company/models/stock_move_line.py

```diff
@@ -0,0 +1,114 @@
+        for move_line, lot_name in moves_to_sync.items():
+            move_line._sync_intercompany_move(lot_name, vals)
+        return res
+
+    def _sync_intercompany_move(self, lot_name, vals):
+        """
+        Sync the intercompany stock move lines
+        with the changes made in this move line.
+        """
+        self.ensure_one()
+        fields_to_sync = self._get_fields_to_sync_intercompany()
+        po_moves = self._get_stock_moves_to_sync()
+        for po_move in po_moves:
+            po_move_line = po_move.move_line_ids.filtered(
+                lambda x: not self.lot_id or x.lot_id.name == lot_name
+            )
+            if not po_move_line:
+                continue
+            vals_to_write = {}
+            for field in fields_to_sync:
+                if field not in vals:
+                    continue
+                field_value = self[field]
+                if field == "lot_id" and field_value:
+                    dest_lot = self._get_or_create_lot_intercompany(
+                        po_move_line.company_id
+                    )
+                    field_value = dest_lot.id
+                vals_to_write[field] = field_value
+            if vals_to_write:
+                po_move_line.write(vals_to_write)
+
+    def _get_stock_moves_to_sync(self):
+        """
+        Get the stock moves that need to be synced with the intercompany move.
+        """
+        self.ensure_one()
+        purchase_line_origin = self.move_id.sale_line_id.sudo().auto_purchase_line_id
+        po_moves = self.env["stock.move"]
+        if purchase_line_origin:
+            po_moves = purchase_line_origin.move_ids.filtered(
+                lambda m: m.picking_id
+                == self.move_id.picking_id.intercompany_picking_id
+                and m.product_id == self.product_id
+            )
+        return po_moves
+
+    @api.model
+    def _get_fields_to_sync_intercompany(self):
+        """
+        Get the fields that need to be synced with the intercompany move line.
+        :return: set of field names
+        """
+        return {"quantity", "lot_id"}
+
+    def _get_or_create_lot_intercompany(self, dest_company):
+        # search if the same lot exists in destination company
+        self.ensure_one()
+        StockLot = self.env["stock.lot"].sudo()
+        lot = self.lot_id.sudo()
+        if not lot.company_id:
+            # if the lot is global, no need to create another one
+            return lot
+        dest_lot = StockLot.search(
+            [
+                ("product_id", "=", lot.product_id.id),
+                ("name", "=", lot.name),
+                ("company_id", "in", [False] + [dest_company.id]),
+            ],
+            limit=1,
+        )
+        if not dest_lot:
+            # if it doesn't exist, create it by copying from original company
+            dest_lot = lot.copy({"company_id": dest_company.id, "name": lot.name})
+        return dest_lot
```

Comment:
It makes no sense IMO to look for other lot in the counterpart, or copy it.

```suggestion
        if lot.company_id:
            lot.company_id = False
        return lot
```

Reply:
What about this?

Reply:
I’m not sure. You can have a product with a company, so in this case, the lot cannot be shared between companies. See https://github.com/odoo/odoo/blob/224aec2308bb7725bbba238037ba874c393a2e3b/addons/stock/models/stock_lot.py#L135
Additionally, what happens with databases where the lot already has a company set?

Reply:
If the product has company, then you can't use that product in the other company as well!! And yes, if the lot has company, we remove it because it's a lot for an inter-company operation, so it's legit to take that path.

Reply:
IMO it makes sense to force the lot available to other companies and as @pedrobaeza pointed out if the product is not allowed in both companies the next error will be raised.
```
Invalid Operation
Incompatible companies on records:
- “S00028/Stock>Customers” belongs to company “C2 - My Test Company” and “Product” (product_id: 'product by lot C1') belongs to another company.
```

Reply:
I’ve updated the code to ensure that the lot used does not have a company_id set, and I’ve also updated the test.